rfcs/2021-02-04-5802-event-metadata.md

```diff
@@ -0,0 +1,206 @@
+# RFC 5802 - 2021-02-04 - Event Metadata
+
+This RFC introduces a plan to associate persistent metadata with every Vector event, both logs and metrics.
+
+## Scope
+
+This RFC will cover the placement of the metadata, its handling, and
+some potential contents. It will not comprehensively cover the specific
+contents beyond what is required for the initial task, as that will
+depend on implement requirements of the relevant features.
+
+## Motivation
+
+There are two upcoming features planned for Vector that require events
+to be associated with additional data that must persist throughout the
+life of an event:
+
+1. End-to-end latency measurement
+2. End-to-end acknowledgements
+
+Both of these features require each event to be tagged with several
+pieces of data. These datum are not part of the data read from the
+source, nor part of what is sent to the sinks, but exist outside of the
+normal stream.
+
+## Internal Proposal
+
+### Data Structures
+
+The event metadata will be introduced as a new structure named
+`EventMetadata`. This will contain fixed data elements, starting with
+just the ingestion timestamp (which is distinct from the timestamp in
+the event data) but will eventually be extended to contain the source
+name, record identifiers, and possibly acknowledgement status:
+
+```rust
+struct EventMetadata {
+    first_timestamp: DateTime,
+    source_events: SmallVec<SourceEvent>,
+    status: EventDeliveryStatus,
+}
+
+struct SourceEvent {
+    timestamp: DateTime,
+    source_name: String,
+    event_id: String,
+}
+
+enum EventDeliveryStatus {
+    …TBD…
+}
+```
+
+To accommodate the metadata, the current `Event` enum will be renamed to
+`EventData`, and a wrapper structure will replace that name:
+
+```rust
+struct Event {
+    data: EventData,
+    metadata: EventMetadata,
+}
+
+enum EventData {
+    Log(LogEvent),
+    Metric(Metric),
+}
+```
+
+All of the methods on `EventData` will remain in place, but this will
+require additional handling for both the creation of new events (since
+the metadata will need to be filled in) and the "into" methods (`fn
+into_log` and `fn into_metric`). Implementations for `Deref` and
+`DerefMut` will simplify porting existing code to the new structure.
+
+### Vector Protocol
+
+The event protocol is used for both the transmission of events between
+Vector instances and the storage of events in disk buffers. This
+metadata is only relevant within a single instance of Vector, so should
+not be transmitted across the wire. However, for the latter, the
+protobuf definition will need to be expanded to include the metadata by
+adding the appropriate structures.
+
+```protobuf
+message EventWithMetadata {
```

Comment:
It seems like this will be a breaking change? Is there a way we could preserve compatibility between vector versions when this is added?

Reply:
I think there will be a way to set up the messages (ie using different field IDs) such that new events will always fail to parse as old events and vice versa. Then new Vector will try to parse the buffer as a new event first, and fall back to the old.

I don't think it makes sense to send the metadata over the wire, so this only affects the disk buffers.

Reply:
It's possible to add `EventMetadata metadata` field to `EventWrapper` which will be backwards compatible
```
message EventWrapper {
  oneof event {
    Log log = 1;
    Metric metric = 2;
  }
  EventMetadata metadata = 3;
}
```
`metadata` field is by default `Option` so it covers the old and the new form. Old Vector will on encountering unknown field skip it, and new Vector will will assume `None` if the field isn't present.

 Then when building `proto::EventWrapper` for sending over wire, metadata field can just be skipped.

One edge case is when new Vector reads `None` for metadate from a buffer. What should it do? Create new metadata and leave `source_events` empty?

Reply:
Good points. I forget that these fields in protobuf are optional. I will adjust the RFC as you suggest.

I think Vector will have to put in some stubs when reading null metadata from a buffer. That is, set the timestamp to now and the source to some dummy field indicating we don't know the original source.